File: zenoh-grpc-client-sdk/zenoh-grpc-cpp/examples/querier.cpp
#include <iostream>

#include "common.hpp"

int main(int argc, char** argv) {
    auto session = zenoh_grpc::Session::connect(example_endpoint(argc, argv));

    zenoh_grpc::Session::QuerierOptions options;
    options.consolidation = zenoh_grpc::ConsolidationMode::None;
    options.timeout_ms = 3000;
    auto querier = session.declare_querier("demo/query/**", options);

    zenoh_grpc::Querier::GetOptions get_options;
    get_options.payload = std::vector<std::uint8_t>{'h', 'a', 'h', 'a', 'h', 'a', 'h', 'a'};
    get_options.encoding = "text/plain";

    auto replies = querier.get_stream("", get_options);
    std::cout << "query sent, waiting for replies..." << std::endl;
    for (const auto& reply : replies) {
        print_reply(reply);
    }

    std::cout << "reply stream dropped=" << replies.dropped_count() << std::endl;
    return 0;
}

File: zenoh-grpc-client-sdk/zenoh-grpc-python/examples/querier.py
import zenoh_grpc


with zenoh_grpc.Session.connect() as session:
    querier = session.declare_querier(
        "demo/query/**", 
        consolidation=zenoh_grpc.ConsolidationMode.NONE, # optional, default is AUTO(LATEST)
        timeout_ms=3_000) # optional, default is 3000ms. The timeout for the queryables to reply. 
    # The querier will receive the replies one by one or all at once depending on the consolidation mode set by the querier (AUTO(default,LATEST), NONE(receive one by one), MONOTONIC(time monotonically consolidation), LATEST(only the latest)).

    replies = querier.get(payload=b"hahahaha", 
                          encoding="text/plain") 
    # the encoding is just a hint for the queryable to know how to parse the payload, it does not affect the actual payload sent to the queryable.
    # unblocking, returns a ReplyStream

    print("query sent, waiting for replies...")

    for reply in replies:
        if reply.ok:
            print("sample:", reply.sample.key_expr, reply.sample.payload, reply.sample.encoding)
        else:
            print("error:", reply.error.payload, reply.error.encoding)
